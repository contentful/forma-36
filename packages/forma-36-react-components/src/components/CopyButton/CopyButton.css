@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';
@import 'settings/borders';

.CopyButton {
  display: inline-block;
  position: relative;
}

.CopyButton__button {
  composes: focus-outline--default from './../../styles/settings/a11y.css';
  height: calc(1rem * (40 / var(--font-base-default)));
  width: calc(1rem * (40 / var(--font-base-default)));
  border: 1px solid var(--color-element-mid);
  background: var(--color-white);
  flex-basis: 100%;
  transition: background var(--transition-duration-short)
    var(--transition-easing-default);
  padding: 0;

  &:hover {
    cursor: pointer;
    background-color: var(--color-element-lightest);
  }

  &:active {
    transition: none;
    background-color: var(--color-element-light);
  }

  &:focus {
    box-shadow: var(--glow-muted);
  }
  & > svg {
    margin-top: calc(1rem * (3 / var(--font-base-default)));
  }
}

.CopyButton__text {
  composes: sr-only from './../../styles/settings/helpers.css';
}

.CopyButton .CopyButton__TabFocusTrap {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
}
