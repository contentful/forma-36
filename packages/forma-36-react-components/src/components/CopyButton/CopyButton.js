import React from 'react';
import CopyToClipboard from 'react-copy-to-clipboard';
import PropTypes from 'prop-types';
import cn from 'classnames';
import Icon from '../Icon';
import Tooltip from '../Tooltip';
import TabFocusTrap from '../TabFocusTrap';
import styles from './CopyButton.css';

export class CopyButton extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    testId: PropTypes.string,
    copyValue: PropTypes.string,
    onCopy: PropTypes.func,
  };

  static defaultProps = {
    extraClassNames: undefined,
    copyValue: undefined,
    testId: 'cf-ui-copy-button',
    onCopy: () => {},
  };

  state = {
    copied: false,
  };

  onCopy = e => {
    this.props.onCopy(e);
    this.setState({ copied: true });
    setTimeout(() => {
      this.setState({ copied: false });
      this.copyButton.blur();
    }, 1000);
  };

  render() {
    const { copyValue, extraClassNames, testId, ...otherProps } = this.props;

    const classNames = cn(styles.CopyButton, extraClassNames);

    return (
      <div
        ref={ref => {
          this.tooltipAnchor = ref;
        }}
        className={classNames}
        id="copyButton"
        data-test-id={testId}
        {...otherProps}
      >
        <CopyToClipboard text={copyValue} onCopy={this.onCopy}>
          <Tooltip
            content={
              this.state.copied ? (
                'Copied!'
              ) : (
                <span>
                  Copy to <br /> clipboard
                </span>
              )
            }
          >
            <button
              type="button"
              ref={ref => {
                this.copyButton = ref;
              }}
              className={styles.CopyButton__button}
            >
              <TabFocusTrap extraClassNames={styles.CopyButton__TabFocusTrap}>
                <span className={styles.CopyButton__text}>
                  Copy {copyValue} to clipboard
                </span>
                <Icon icon="Copy" color="muted" />
              </TabFocusTrap>
            </button>
          </Tooltip>
        </CopyToClipboard>
      </div>
    );
  }
}

export default CopyButton;
