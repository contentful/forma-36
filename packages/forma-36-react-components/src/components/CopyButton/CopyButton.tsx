import React, { useCallback, useRef, useState } from 'react';
import CopyToClipboard from 'react-copy-to-clipboard';

import cn from 'classnames';
import Icon from '../Icon';
import { Tooltip } from '../Tooltip';
import type { TooltipPlace } from '../Tooltip';
import TabFocusTrap from '../TabFocusTrap';
import styles from './CopyButton.css';

export interface CopyButtonProps {
  copyValue?: string;
  onCopy?: (value: string) => void;
  className?: string;
  testId?: string;
  tooltipPlace?: TooltipPlace;
  tooltipText?: React.ReactNode;
  tooltipCopiedText?: React.ReactNode;
}

export interface CopyButtonState {
  copied: boolean;
}

export function CopyButton({
  copyValue,
  className,
  testId,
  onCopy,
  tooltipPlace,
  tooltipText,
  tooltipCopiedText,
  ...otherProps
}: CopyButtonProps) {
  const [copied, setCopied] = useState<boolean>(false);
  const button = useRef<HTMLButtonElement | null>(null);
  const tooltipAnchor = useRef<HTMLDivElement | null>(null);

  const handleCopy = useCallback(
    (e: string) => {
      if (onCopy) {
        onCopy(e);
      }

      setCopied(true);

      setTimeout(() => {
        setCopied(false);
        if (button.current) {
          button.current.blur();
        }
      }, 1000);
    },
    [onCopy],
  );

  const classNames = cn(styles['CopyButton'], className);

  return (
    <div
      ref={tooltipAnchor}
      className={classNames}
      id="copyButton"
      data-test-id={testId}
      {...otherProps}
    >
      <CopyToClipboard text={copyValue || ''} onCopy={handleCopy}>
        <Tooltip
          place={tooltipPlace}
          content={
            copied ? (
              <span>{tooltipCopiedText}</span>
            ) : (
              <span>{tooltipText}</span>
            )
          }
        >
          <button
            type="button"
            ref={button}
            className={styles['CopyButton__button']}
          >
            <TabFocusTrap className={styles['CopyButton__TabFocusTrap']}>
              <span className={styles['CopyButton__text']}>
                Copy {copyValue} to clipboard
              </span>
              <Icon icon="Copy" color="muted" />
            </TabFocusTrap>
          </button>
        </Tooltip>
      </CopyToClipboard>
    </div>
  );
}

CopyButton.defaultProps = {
  testId: 'cf-ui-copy-button',
  tooltipText: (
    <React.Fragment>
      Copy to <br /> clipboard
    </React.Fragment>
  ),
  tooltipCopiedText: <React.Fragment>Copied!</React.Fragment>,
};

export default CopyButton;
