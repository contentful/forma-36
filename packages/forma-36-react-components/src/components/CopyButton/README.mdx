# CopyButton

The CopyButton is a styled button that copies text into the user's clipboard. This is useful in combination with for example tables or inputs that show values that the user is likely to use elsewhere, such as automatically generated passwords.

## Table of contents

- [How to use CopyButton](#how-to-use-CopyButton)
- [Code examples](#code-examples)
- [Content recommendations](#content-recommendations)
- [Accessibility](#accessibility)

## How to use CopyButton

CopyButton can be used in combination with text inputs where the input value should be easy to copy. In such cases it is best to position the CopyButton at the end of the input.

## Code examples

```jsx
// import { CopyButton } from '@contentful/forma-36-react-components';

<CopyButton copyValue="Lorem Ipsum" />
```

An example with a custom tooltip:

```jsx
// import { CopyButton } from '@contentful/forma-36-react-components';

<CopyButton
  copyValue="Lorem Ipsum"
  tooltipCopiedText="The text has been copied"
  tooltipText="Copy this text to clipboard"
/>
```
