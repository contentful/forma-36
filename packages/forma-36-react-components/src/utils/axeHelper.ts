import { configureAxe } from 'jest-axe';

const axe = configureAxe({
  rules: {
    region: { enabled: false },
  },
});

export default axe;
