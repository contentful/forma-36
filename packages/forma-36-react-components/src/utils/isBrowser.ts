const isBrowser = () => typeof window !== 'undefined';

export default isBrowser;
