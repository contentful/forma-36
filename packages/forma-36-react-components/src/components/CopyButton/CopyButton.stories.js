import React from 'react';
import { storiesOf } from '@storybook/react';
import { text } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';

import CopyButton from './CopyButton';

storiesOf('Components|CopyButton', module)
  .addDecorator(withInfo)
  .add('default', () => (
    <div>
      <CopyButton
        extraClassNames={text('Extra Class Names', '')}
        copyValue={text('Copy Value', 'Lorem Ipsum')}
      />
    </div>
  ));
