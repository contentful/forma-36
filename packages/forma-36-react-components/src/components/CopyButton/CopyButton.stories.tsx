import React from 'react';
import type { Meta, Story } from '@storybook/react/types-6-0';

import CopyButton from './CopyButton';
import type { CopyButtonProps } from './CopyButton';
import notes from './README.mdx';

export default {
  argTypes: {
    children: { control: { disable: true } },
    className: { control: { disable: true } },
    testId: { control: { disable: true } },
    tooltipText: { control: 'text' },
    tooltipCopiedText: { control: 'text' },
  },
  component: CopyButton,
  parameters: {
    propTypes: CopyButton['__docgenInfo'],
    notes,
  },
  title: 'Components/CopyButton',
} as Meta;

export const Default: Story<CopyButtonProps> = (args) => {
  return <CopyButton {...args} />;
};

Default.args = {
  copyValue: 'Lorem Ipsum',
  tooltipCopiedText: 'Copied!',
  tooltipPlace: 'bottom',
  tooltipText: 'Copy to clipboard',
};
