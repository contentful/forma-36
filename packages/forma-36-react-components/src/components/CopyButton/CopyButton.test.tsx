import React from 'react';
import { render } from '@testing-library/react';

import { axe } from '../../utils/axeHelper';
import { CopyButton } from './CopyButton';

it('renders the component', () => {
  const { container } = render(<CopyButton />);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const { container } = render(<CopyButton className="my-extra-class" />);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with a copy value', () => {
  const { container } = render(<CopyButton copyValue="test" />);

  expect(container.firstChild).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const { container } = render(<CopyButton />);
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});
