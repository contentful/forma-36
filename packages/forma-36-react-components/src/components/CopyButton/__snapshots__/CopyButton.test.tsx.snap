// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  class="CopyButton"
  data-test-id="cf-ui-copy-button"
  id="copyButton"
>
  <span>
    <button
      class="CopyButton__button"
      type="button"
    >
      <span
        class="TabFocusTrap CopyButton__TabFocusTrap"
        tabindex="-1"
      >
        <span
          class="CopyButton__text"
        >
          Copy 
           to clipboard
        </span>
        <svg
          class="Icon Icon--small Icon--muted"
          data-test-id="cf-ui-icon"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
          />
        </svg>
      </span>
    </button>
  </span>
</div>
`;

exports[`renders the component with a copy value 1`] = `
<div
  class="CopyButton"
  data-test-id="cf-ui-copy-button"
  id="copyButton"
>
  <span>
    <button
      class="CopyButton__button"
      type="button"
    >
      <span
        class="TabFocusTrap CopyButton__TabFocusTrap"
        tabindex="-1"
      >
        <span
          class="CopyButton__text"
        >
          Copy 
          test
           to clipboard
        </span>
        <svg
          class="Icon Icon--small Icon--muted"
          data-test-id="cf-ui-icon"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
          />
        </svg>
      </span>
    </button>
  </span>
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<div
  class="CopyButton my-extra-class"
  data-test-id="cf-ui-copy-button"
  id="copyButton"
>
  <span>
    <button
      class="CopyButton__button"
      type="button"
    >
      <span
        class="TabFocusTrap CopyButton__TabFocusTrap"
        tabindex="-1"
      >
        <span
          class="CopyButton__text"
        >
          Copy 
           to clipboard
        </span>
        <svg
          class="Icon Icon--small Icon--muted"
          data-test-id="cf-ui-icon"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
          />
        </svg>
      </span>
    </button>
  </span>
</div>
`;
